app/controllers/application_controller.rb

```diff
@@ -63,4 +64,21 @@ def store_user_location!
     # :user is the scope we are authenticating
     store_location_for(:user, request.fullpath)
   end
+
+  def can_redirect_to(redirect_to) # rubocop:disable Metrics/MethodLength
+    return unless redirect_to
+
+    hostname = begin
+      URI.parse(redirect_to).hostname
+    rescue URI::InvalidURIError
+      nil
+    end
+    return unless hostname
+
+    apps = Application.arel_table
+    return redirect_to if Application.where(apps[:redirect_uri].matches("%#{hostname}%")).any? ||
```

Comment:
what does this `matches("%#{hostname}%")` mean?

Reply:
is this a regex?
if so, it should be delimited

Reply:
It takes advantage of Arel (The library that builds Active Record things into SQL) to create a query that looks like `SELECT * FROM applications WHERE "applications"."redirect_uri" ILIKE '%hostname%'` without having to resort to writing the sql directly, and working to ensure that it's correctly escaped

Reply:
https://stackoverflow.com/questions/4430578/how-to-do-a-like-query-in-arel-and-rails#7012743

Reply:
Not a regular expression :+1:

Reply:
would a match on `https://#{hostname}%` be safer?

Reply:
No, because the redirect_uri for a SAML endpoint is a _hostname_ not a URL. The goal with this change is to restrict the redirect_uri so that we don't redirect to random things, but as long as the hostname matches an existing service that we've registered, it should be OK.

The real alternative solution, to me, is to remove the functionality of redirecting to external services completely, but that's non-ideal for users.

Reply:
Slightly more accurate, the redirect_hosts _may_ be a hostname, or a full URI

Reply:
I was misunderstanding specifically which part this question was referring to, which is the Application, rather than the SamlServiceProvider.